File: frontend/src/api/normalization.ts
import { governedFetch } from "./context";
import { API_BASE_URL } from "./health";
import type { Artifact, PipelineStep } from "./ingestion";

export interface NormalizationSummary { id: number; tenant_id: number; source_file_id: number; ingestion_run_id: number; status: string; started_at: string; completed_at: string | null; staging_count: number; canonical_count: number; exception_count: number; mapping_code: string | null; mapping_version: string | null; normalization_version: string | null; no_op: boolean; error_message: string | null; steps: PipelineStep[]; artifacts: Artifact[]; }
export interface NormalizationControl { id: number; control_name: string; staging_value: string | null; canonical_value: string | null; difference_value: string | null; status: string; }
export interface NormalizationException { id: number; staging_entity_type: string; staging_entity_id: number; exception_code: string; exception_type: string; severity: string; field_name: string | null; observed_value: string | null; expected_value: string | null; message: string; status: string; created_at: string; }
interface Page<T> { items: T[]; total: number; page: number; page_size: number; }
async function request<T>(path: string, init?: RequestInit): Promise<T> { const response = await governedFetch(`${API_BASE_URL}${path}`, init); const body = (await response.json()) as T & { detail?: string }; if (!response.ok) throw new Error(body.detail ?? `Request failed (${response.status})`); return body; }
export function normalizeIngestion(ingestionRunId: number, mappingCode?: string, forceRerun = false) { return request<NormalizationSummary>(`/ingestions/${ingestionRunId}/normalize`, { method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify({ mapping_code: mappingCode || null, force_rerun: forceRerun }) }); }
export async function fetchNormalizations(ingestionRunId: number) { return (await request<Page<NormalizationSummary>>(`/ingestions/${ingestionRunId}/normalizations`)).items; }
export function fetchNormalizationControls(runId: number) { return request<NormalizationControl[]>(`/normalizations/${runId}/control-totals`); }
export async function fetchNormalizationExceptions(runId: number) { return (await request<Page<NormalizationException>>(`/normalizations/${runId}/exceptions?page_size=200`)).items; }
export async function fetchCanonical(type: string) { return (await request<Page<Record<string, unknown>>>(`/canonical/${type}?page_size=100`)).items; }
export async function fetchCanonicalLineage(type: string, id: number) { return (await request<Page<Record<string, unknown>>>(`/canonical/${type}/${id}/lineage`)).items; }

File: backend/app/schemas/normalization.py
from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, ConfigDict

from app.schemas.ingestion import ArtifactResponse, StepResponse


class NormalizeRequest(BaseModel):
    mapping_code: str | None = None
    force_rerun: bool = False


class NormalizationSummary(BaseModel):
    id: int
    tenant_id: int
    source_file_id: int | None
    ingestion_run_id: int | None
    status: str
    started_at: datetime
    completed_at: datetime | None
    staging_count: int
    canonical_count: int
    exception_count: int
    mapping_code: str | None
    mapping_version: str | None
    normalization_version: str | None
    no_op: bool = False
    error_message: str | None
    steps: list[StepResponse] = []
    artifacts: list[ArtifactResponse] = []


class NormalizationExceptionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    pipeline_run_id: int
    source_file_id: int
    staging_entity_type: str
    staging_entity_id: int
    exception_code: str
    exception_type: str
    severity: str
    field_name: str | None
    observed_value: str | None
    expected_value: str | None
    message: str
    status: str
    exception_fingerprint: str
    metadata_json: dict[str, Any] | None
    created_at: datetime
    updated_at: datetime


class NormalizationControlResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    pipeline_run_id: int
    source_file_id: int
    control_name: str
    staging_value: Decimal | None
    canonical_value: Decimal | None
    difference_value: Decimal | None
    tolerance: Decimal | None
    status: str
    metadata_json: dict[str, Any] | None
